src/runtime/hexagon_remote/qurt_thread_pool.cpp

```diff
@@ -0,0 +1,271 @@
+struct work_queue_t {
+    // all fields are protected by this mutex.
+    qurt_mutex_t work_mutex;
+
+    // Jobs that the thread pool needs to work on.
+    work *jobs;
+
+    // Global flag indicating if the thread pool has been shutdown.
+    bool shutdown;
+
+    bool running() {
+        return !shutdown;
+    }
+    void lock() {
+        qurt_mutex_lock(&work_mutex);
+    }
+    void unlock() {
+        qurt_mutex_unlock(&work_mutex);
+    }
+    void init() {
+        qurt_mutex_init(&work_mutex);
+        lock();
+        shutdown = false;
+        jobs = NULL;
+        unlock();
+    }
+};
+work_queue_t work_queue;
+qurt_thread_t threads[MAX_WORKER_THREADS];
+
+// This function does the real work of the thread pool.
+// owned_job is used to tell the differece between the master thread
+// and the worker threads. If owned_job is non-null then it means that
+// we are in the master thread, i.e. the thread that put owned_job in
+// the work queue. This thread, like other worker threads in this function,
+// tries to acquire a lock on the work queue. Each thread then looks for
+// a job to do. In the case of the master thread, it is "owned_job" while
+// a worker thread has to iterate backwards over the work queue and find
+// a job to work on. Once a job has been found, the thread, after some
+// book-keeping, releases the lock and call halide_do_task to do the job.
+// After the job is done, the lock on the queue is re-acquired to update
+// the status and the thread loops again to look for new work. If a worker
+// thread doesn't find any new work, it goes to sleep until awoken by
+// the master thread.
+void worker_thread(work *owned_job) {
+    // You can work only if you get a lock on the work queue.
+    unsigned int tid = qurt_thread_get_id();
+
+    bool locked = owned_job != NULL;
+
+    // ***********************
+    // *** Lock work queue ***
+    // ***********************
+    work_queue.lock();
+
+    // If I'm a job owner, then I was the thread that called
+    // do_par_for, and I should only stay in this function while there
+    // are active workers on the job (i.e. running()) or if the job
+    // hasn't been claimed entirely (!claimed()). If I'm a lowly worker
+    // thread, I should stay in this function as long as the work queue is running.
+    while (owned_job != NULL ? (owned_job->running() || !owned_job->claimed())
+           : work_queue.running()) {
+
+        work *job = NULL;
+        if (!owned_job) {
+            // If this threads doesn't own a job, it looks for one
+            // and tries to do it. If it cannot find a job, it goes
+            // to sleep.
+            job = work_queue.jobs;
+            if (!job) {
+                // release the lock and go to sleep.
+                // ***************************
+                // *** Work queue unlocked ***
+                // ***************************
+                work_queue.unlock();
+                qurt_sem_down(&wait_for_work);
```

Comment:
I'd like to note that the use of qurt_sem_down is unfortunate. While it does work fine for our needs, I'd like to change it to use qurt_cond_wait. However, given my priorities currently, I'll fix this once this has been merged to master.